clippy_lints/src/if_let_redundant_pattern_matching.rs

```diff
@@ -0,0 +1,81 @@
+use rustc::lint::*;
+use rustc::hir::*;
+use utils::{paths, span_lint_and_then, match_path, snippet};
+
+/// **What it does:*** Lint for redundant pattern matching over `Result` or `Option`
+///
+/// **Why is this bad?** It's more concise and clear to just use the proper utility function
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rustc
+/// if let Ok(_) = Ok::<i32, i32>(42) {}
+/// if let Err(_) = Err::<i32, i32>(42) {}
+/// if let None = None::<()> {}
+/// if let Some(_) = Some(42) {}
+/// ```
+/// The more idiomatic use would be:
+/// ```rustc
+/// if Ok::<i32, i32>(42).is_ok() {}
+/// if Err::<i32, i32>(42).is_err() {}
+/// if None::<()>.is_none() {}
+/// if Some(42).is_some() {}
+/// ```
+declare_lint! {
+    pub IF_LET_REDUNDANT_PATTERN_MATCHING,
+    Warn,
+    "use the proper utility function avoiding an `if let`"
+}
+
+#[derive(Copy, Clone)]
+pub struct Pass;
+
+impl LintPass for Pass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(IF_LET_REDUNDANT_PATTERN_MATCHING)
+    }
+}
+
+impl LateLintPass for Pass {
+    fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
+
+        if let ExprMatch(ref op, ref arms, MatchSource::IfLetDesugar{..}) = expr.node {
+
+            if arms[0].pats.len() == 1 {
+
+                let good_method = match arms[0].pats[0].node {
+                    PatKind::TupleStruct(ref path, ref pats, _) if pats.len() == 1 && pats[0].node == PatKind::Wild =>  {
+
+                        if match_path(path, &paths::RESULT_OK) {
+                            "is_ok()"
+                        } else if match_path(path, &paths::RESULT_ERR) {
+                            "is_err()"
+                        } else if match_path(path, &paths::OPTION_SOME) {
+                            "is_some()"
+                        } else {
+                            return
+                        }
+                    }
+
+                    PatKind::Path(_, ref path) if match_path(path, &paths::OPTION_NONE) => {
+                        "is_none()"
+                    }
+
+                    _ => return
+                };
+
+                span_lint_and_then(cx,
+                                   IF_LET_REDUNDANT_PATTERN_MATCHING,
+                                   arms[0].pats[0].span,
+                                   &format!("redundant pattern matching, consider using `{}`", good_method),
+                                   |db| {
+                                        db.span_suggestion(expr.span,
+                                                           "try this",
+                                                           format!("{}.{}", snippet(cx, op.span, "_"), good_method));
+                                    });
```

Comment:
actually I did it that way intentionally, because I think that having all the body and the else is a bit noisy, isn't it? I'd like to keep it obvious that the error is in the if let. I could change the line in `format!("if {}.{} {{", snippet(cx, op.span, "_"), good_method));` if that's more helpful.